tutorials/haskell/golden-testing-in-haskell-with-hspec-gold/tutorial.md

```diff
@@ -0,0 +1,190 @@
+with some additional libs to implement our own Golden type.
+
+
+###Writing our first golden test
+The [hspec-golden](https://github.com/stackbuilders/hspec-golden) library provides us with an easy example to test the functionality of the library with the `defaultGolden` helper:
+
+``` 
+describe "id" $
+     it "generates the right output with the right params" $
+        let output = show $ id (1:: Int)
+        in defaultGolden "id" output
+```
+
+On this test we are setting up a B.D.D. test with [hspec](https://github.com/hspec/hspec) “describe” and “it”,
+but on the inside there is an implementation of the `defaultGolden` helper,
+that will create(or update) for us a directory called `.golden` this will contain another directory named after our described test in this case `id` with two files containing the `actual` and `golden` values for our test.
+
+For our example we will create our own Golden type.
+We have created a file called test.yml that contains an ansible role on the dir ["test/fixtures/"](https://github.com/MedAdrian/hspec-golden-tutorial/blob/master/test/fixtures/test.yml) 
+we will use this file contents and transform them into JSON.
+
+```
+{--
+Here we define our Golden test as follows:
+- We read a .yml file that contains an ansible role
+- We decode the string read by Haskell that we read form the .yml file
+--}
+
+spec :: Spec
+spec = do
+    describe "conversion" $ do
+      content <- runIO $ B.readFile "test/fixtures/test.yml"
+      let eYaml = Y.decodeEither' content :: Either Y.ParseException Y.Value  --Parses to a generic yml type.
+          Right json = A.encode <$> eYaml                                     --Parses the data type to a json structure.
+      it "transforms a YAML file contents to JSON" $ do
+        myGoldenTest "conversion" json                                        --Generates the golden test.
+```
+
+To see the full code example access this [rerpository](https://github.com/MedAdrian/hspec-golden-tutorial).
+
+If you are confused about how test are written you can follow this [link](http://hspec.github.io/hspec-discover.html).
+
+#####I.e: Directory Structure 
+``` 
+|_.golden
+   |_id
+     |_actual
+     |_golden
+```
+
+For this example we are testing that the output from the function will return the expected golden value.
+
+##Installing the CLI
+You can install the [hspec-golden](https://github.com/stackbuilders/hspec-golden) command line interface (CLI) with stack:
+
+``` $ stack install hspec-golden ```
+
+or cabal:
+
+``` $ cabal install hspec-golden ```
+
+The CLI is called hgold:
+```
+$ hgold
+
+Parameters:
+  DIR    The testing directory where you're dumping your results (default: .golden/)
+
+Flags:
+  -u[DIR]  --update[=DIR]  Replaces `golden` files with `actual` files
+  -v       --version       Displays the version of hgold
+  -h       --help          Displays help information
```

Comment:
@MedAdrian I'm not sure how valuable is the output of the command here. Also, I think the CLI could [change soon](https://github.com/stackbuilders/hspec-golden/pull/8). In fact, I'm looking for someone to help me with that issue. If you have some free time, would you like to take it? If you don't it's fine. :wink:

Reply:
Sure no problem, then i will only add as part of this section that the lib also includes a CLI.